source/spawn/spawn-channels.adb

```diff
@@ -0,0 +1,923 @@
+
+   procedure Shutdown
+     (Parent : in out Glib.IOChannel.Giochannel;
+      Event  : in out Glib.Main.G_Source_Id;
+      Lock   : in out Glib.Gboolean);
+   --  Common code to stop watching and to shutdown IO channel.
+
+   procedure Start_Watch
+     (Parent    : Glib.IOChannel.Giochannel;
+      Event     : in out Glib.Main.G_Source_Id;
+      Lock      : in out Glib.Gboolean;
+      Condition : Glib.IOChannel.GIOCondition;
+      Callback  : access function
+        (Source    : Glib.IOChannel.Giochannel;
+         Condition : Glib.IOChannel.GIOCondition;
+         Data      : access Spawn.Internal.Process_Reference)
+         return Glib.Gboolean;
+      Data      : access Spawn.Internal.Process_Reference)
+     with Convention => C;
+   --  Common code to start (continue) watching of the IO channel.
+
+   ------------------
+   -- Child_Stderr --
+   ------------------
+
+   function Child_Stderr (Self : Channels) return Glib.Gint is
+   begin
+      return Self.Stderr_Child;
+   end Child_Stderr;
+
+   -----------------
+   -- Child_Stdin --
+   -----------------
+
+   function Child_Stdin (Self : Channels) return Glib.Gint is
+   begin
+      return
+        (if Self.Stdin_Child /= -1
+           then Self.Stdin_Child else Self.PTY_Slave);
+   end Child_Stdin;
+
+   ------------------
+   -- Child_Stdout --
+   ------------------
+
+   function Child_Stdout (Self : Channels) return Glib.Gint is
+   begin
+      return
+        (if Self.Stdout_Child /= -1
+           then Self.Stdout_Child else Self.PTY_Slave);
+   end Child_Stdout;
+
+   -----------------------------
+   -- Close_Child_Descriptors --
+   -----------------------------
+
+   procedure Close_Child_Descriptors (Self : in out Channels) is
+      Ignore : Interfaces.C.int;
+
+   begin
+      if Self.Stdin_Child /= -1 then
+         Ignore := Spawn.Posix.close (Interfaces.C.int (Self.Stdin_Child));
+         Self.Stdin_Child := -1;
+      end if;
+
+      if Self.Stdout_Child /= -1 then
+         Ignore := Spawn.Posix.close (Interfaces.C.int (Self.Stdout_Child));
+         Self.Stdout_Child := -1;
+      end if;
+
+      if Self.Stderr_Child /= -1 then
+         Ignore := Spawn.Posix.close (Interfaces.C.int (Self.Stderr_Child));
+         Self.Stderr_Child := -1;
+      end if;
+
+      if Self.PTY_Slave /= -1 then
+         Ignore := Spawn.Posix.close (Interfaces.C.int (Self.PTY_Slave));
+         Self.PTY_Slave := -1;
+      end if;
+   end Close_Child_Descriptors;
+
+   ---------------------
+   -- On_Stderr_Event --
+   ---------------------
+
+   function On_Stderr_Event
+     (source    : Glib.IOChannel.Giochannel;
+      condition : Glib.IOChannel.GIOCondition;
+      data      : access Internal.Process_Reference) return Glib.Gboolean
+   is
+      pragma Unreferenced (source);
+      pragma Unreferenced (condition);
+
+      Self : Channels renames data.Self.Channels;
+
+   begin
+      Self.Stderr_Lock := @ - 1;
```

Comment:
Incoming value of Lock can't be 0 here; because zero value means there is no waiting for IO.

Value may be more that 1 when nested main loop is running from the callback.

Glib's convention of Gboolean is 0 or non-zero, it exactly maps to the value of the lock.